### ISSUE TITLE
encrypting stored data

### ISSUE THREAD

**Issue by codingfabi** [NGO-DB/ndb-core#144]
PouchDB doenst encryped any stored data except for passwords.
We should consider encrypting the data we store, especially names, phone numbers, adresses and test results.
i found following PouchDB-Plugin:
https://github.com/calvinmetcalf/crypto-pouch

Any ideas or opinions?

**Comment by sleidig** [comment]
Yes, this is an important point! I was also looking at crypto-pouch before.

Related to this is the more far-reaching question of access control - maybe some users will only have rights to see some of the data. If the whole CouchDB/PouchDB database is synced and accessible it is not secure then.
My current idea is to download a crypto-pouch encrypted database for each user, encrypted with the user password, which only contains the data that user has rights to see. On a device there may be several database instances, encrypted with different users' passwords in parallel then.

**Comment by codingfabi** [comment]
Makes sence to me, using the user Password to encrypt is a good idea as well. but it leaves us with 2 tasks:
- we still need to encrypt the centralized database. Use a master Password for encryption as well?
- we need a concept of which user is allowed to see whicht parts of the database.

**Comment by mspoeri** [comment]
You are aware about the fact that encrypted data can not be searched or indexed?

Also I really really hope, that the passwords itself are NOT stored (regardless if encrypted or not) in the database?!?

**Comment by sleidig** [comment]
Without indices our current database design is dead slow. If there are no workarounds maybe we even have to consider switching the database library.

In order to allow full-featured offline use (this is essential for our target users) I also don't see any alternative to storing hashed passwords locally. This is to some extent mitigated by the fact that one needs at least initially correct credentials to download the local database.
Maybe we can think a more secure system though. Possibly something like a pin for local login that is not identical with the password for remote database access needed for syncing. But if the user has to enter two different passwords on login that will be horrible UX. Ideas?

**Comment by codingfabi** [comment]
Let's get this topic back on the horizon: 

For start I was wondering, what the status quo of our data encryption and our security measures in general is. What are we doing so far considering the common safety objectives confidentiality, integrity and availability? And what needs to be done to fulfill those safety objectives? Are there any realisitc attacking scenarios?

Should we consider testing our application with a penetration test software like Arachni or openVAS? Or should we even consider outsourcing a professional test?

**Comment by sleidig** [comment]
+1 for getting the scenarios and requirements clear. Maybe a discussion topic for the weekend meet?

@fschoell was particularly interested in security topics, if I remember right. Fred, can you add your thoughts maybe?

**Comment by sleidig** [comment]
This may be an interesting reference (or basis) for end-to-end client-side encryption: https://docs.mongodb.com/manual/core/security-client-side-encryption/

**Comment by sleidig** [comment]
Implementation planning for this moves to 
- https://github.com/Aam-Digital/internal/issues/355
- also see https://github.com/Aam-Digital/internal/issues/352

(security-related issues are discussed in the internal, private workspace)
